Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx

```diff
@@ -1,24 +1,225 @@
+        type: "changed",
+        field: "speed", // Correctly specify the field
+        payload: value === "", // Set to true if the value is an empty string
+      });
 
+      console.log("Speed:", value); // Add this line
+    },
+    [],
+  );
+
+  const handleBatteryChange = useCallback(
+    (e: React.ChangeEvent<HTMLInputElement>) => {
+      const value = e.target.value;
+      // Allow empty input, or convert to a number
+      //setBattery(value === "" ? "" : Number(value));
+      dispatch({
+        type: "Update battery",
+        payload: value === "" ? "" : Number(value),
+      });
+      console.log("Battery:", value); // Add this line
+      if (value === "") {
+        dispatch({
+          type: "Update valid",
+          payload: false,
+        });
+      }
+    },
+    [],
+  );
+
+  const handleWeatherChange = useCallback(
+    (e: React.ChangeEvent<HTMLInputElement>) => {
+      const value = e.target.value;
+      // Allow empty input, or convert to a number
+      dispatch({
+        type: "Update weather",
+        payload: value === "" ? "" : Number(value),
+      });
+      console.log("weather:", value); // Add this line
+      if (value === "") {
+        dispatch({
+          type: "Update valid",
+          payload: false,
+        });
+      }
+    },
+    [],
+  );
+  const validInputs = useMemo(() => {
+    //reset battery and speed
+    let speedError = "";
+    let batteryError = "";
+    let isValid = true;
+
+    //check the chnaged fields
+    if (state.changed.speed) {
+      if (speed === "") {
+        isValid = false;
+        speedError = "Speed is required";
+      } else if (Number(speed) < 0 || Number(speed) > 90) {
+        isValid = false;
+        speedError = "The speed should be within the range of 0 to 90";
+      }
+    }
+
+    if (state.changed.battery) {
+      if (battery === "") {
+        isValid = false;
+        batteryError = "Battery is required";
+      } else if (
+        typeof battery === "number" &&
+        (battery < 0 || battery > 100)
+      ) {
+        isValid = false;
+        batteryError =
+          "The battery percentage should be within the range of 0 to 100";
+      }
+    }
+
+    return { speedError, batteryError, isValid };
+  }, [state.changed, speed, battery]); //recompute the answer if speed, battery, or weather change
+
+  const handleButtonClick = () => {
+    const { speedError, batteryError, isValid } = validInputs;
+
+    if (!isValid) {
+      console.error(speedError, batteryError);
+      return;
+    }
+    //calculate range and then in the return function, call valid inputs. if it returns false, print null, else print confirmatin message
+    range =
+      -((Number(speed) * Number(speed) * Number(battery)) / 2500) +
+      4 * Number(battery) +
+      Number(weather);
+
+    console.log(`Calculated range: ${range.toFixed(2)} km`);
+
+    // Directly manipulate the DOM to show the calculated range
+    document.getElementById(
```

Comment:
This logic doesn't need to be in a function. we also shouldn't be manipulating the dom here. 

You should memoize the range and make sure it's only recomputed when state.speed, state.battery and state.weather change. Be sure to check that the inputs are valid before recomputing